### PR TITLE
[Experimental] Surface custom fields in My Account pages

### LINKED ISSUE #42779: Surface custom fields in My Account Order page.
Custom fields should be visible in the order screen in My Account page.

It's unclear where additional fields (ones that don't belong to Billing or Shipping) and we might need design input.

image

### PR DESCRIPTION
**Pull request by mikejolley**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Shows checkout custom field data under:

- My Account > View Orders
- My Account > Addresses
- My Account > Addresses > Edit
- My Account > Account details

Supporting changes:

- Introduced `woocommerce_edit_account_form_fields` above the password field
- Updated `woocommerce_form_field` to post a value when using checkboxes even if unchecked
- Introduced `woocommerce_my_account_after_my_address` and `woocommerce_order_details_after_customer_address` hooks
- Updates to styling

Closes #42779

Screenshots:

Screenshot 2024-01-19 at 14 59 35
Screenshot 2024-01-19 at 14 59 44
Screenshot 2024-01-19 at 14 59 50
Screenshot 2024-01-19 at 15 00 01

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Before testing, implement the snippet to add custom fields as found in #43449
2. Place an order using a logged in account, making sure to fill out the custom fields during checkout
3. Go to the my account page
4. In the orders section, view the order you just placed and check that custom field values are displayed
5. In the account details section, confirm the custom fields you registered are visible as form fields and editable. Edit the values and confirm they persist. 
6. In the address section, check custom fields are shown.
7. Edit an address. Confirm the form fields for custom fields are shown, and you can edit a value and have it persist.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [x] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

Moved `woocommerce_edit_account_form` above the password field.
Introduced `woocommerce_my_account_after_my_address` and `woocommerce_order_details_after_customer_address` hooks.

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce/src/Blocks/Domain/Services/CheckoutFieldsFrontend.php
Comment (opr) [inline plugins/woocommerce/src/Blocks/Domain/Services/CheckoutFieldsFrontend.php]: Can you confirm this is fine because nonce verification took place in the core form handling? (Same for other instances of this)

File: plugins/woocommerce/src/Blocks/Domain/Services/CheckoutFieldsFrontend.php
Comment (opr) [inline plugins/woocommerce/src/Blocks/Domain/Services/CheckoutFieldsFrontend.php:252]: Should this also be validating it against any validation hooks in place? e.g. add this code:

```php

woocommerce_blocks_register_checkout_field(
			array(
				'id'       => 'plugin-namespace/gov-id',
				'label'    => 'Government ID',
				'location' => 'address',
				'type'     => 'text',
				'required' => true,
			),
		);

add_filter('woocommerce_blocks_validate_additional_field_plugin-namespace/gov-id', function ( \WP_Error $error, $value, $schema, $request ) {
			$match = preg_match( '/[A-Z0-9]{5}/', $value );
			if ( 0 === $match || false === $match ) {
				$error->add( 'invalid_gov_id', 'Please ensure your government ID matches the correct format.' );
			}
			return $error;
		}, 10, 4);
```

then save your address with an invalid format (e.g. `abcde`) I would expect this not to save. Same for contact

File: plugins/woocommerce/includes/wc-template-functions.php
Comment (opr) [inline plugins/woocommerce/includes/wc-template-functions.php]: should `$args['label']` get `esc_html` ? Doesn't look like `wp_parse_args` does any cleaning?

**Review by opr — CHANGES_REQUESTED** [review]
Thanks Mike this is looking great, displaying, editing, and saving working well. Just a few comments that need to be looked at/confirmed before approving.

File: plugins/woocommerce/src/Blocks/Domain/Services/CheckoutFieldsFrontend.php
Comment (mikejolley) [reply plugins/woocommerce/src/Blocks/Domain/Services/CheckoutFieldsFrontend.php]: Yeah, we're hooking into things like `woocommerce_after_save_address_validation` where nonce was already checked.

File: plugins/woocommerce/src/Blocks/Domain/Services/CheckoutFieldsFrontend.php
Comment (mikejolley) [reply plugins/woocommerce/src/Blocks/Domain/Services/CheckoutFieldsFrontend.php:252]: Could we tackle this in a follow up? The current validation method requires schema so I'm not certain how to consume that from these classes. If we have a method we can call, we can add a validation_callback when registering fields and add support in woo core to check it.

Comment (opr) [reply plugins/woocommerce/src/Blocks/Domain/Services/CheckoutFieldsFrontend.php:252]: Yeah, makes sense to do this as a follow up, any further orders with this info will be re-validated anyway

Comment (opr) [reply plugins/woocommerce/src/Blocks/Domain/Services/CheckoutFieldsFrontend.php:252]: Added an issue: https://github.com/woocommerce/woocommerce/issues/44019

File: plugins/woocommerce/src/Blocks/Domain/Services/CheckoutFieldsFrontend.php
Comment (opr) [inline plugins/woocommerce/src/Blocks/Domain/Services/CheckoutFieldsFrontend.php:191]: It is also failing lint because of the missing nonce verification here

File: plugins/woocommerce/templates/myaccount/form-edit-account.php
Comment (opr) [inline plugins/woocommerce/templates/myaccount/form-edit-account.php]: Lint is complaining that this isn't accompanied by a docblock, can we just add a short description of the hook, even if it wasn't there before it might be nice to put one in.

**Review by opr — APPROVED** [review]
Thanks Mike, looking good. Just two suggestions that shouldn't require a re-review so approving

**Comment by mikejolley** [comment]
@lsinger Could I get your approval on the hook addition and movement in this PR? The move is so any hooked in fields appear above the change password inputs which is a more logical field order inline with designs. 

If thats not acceptable, I can introduce a new hook instead (and would want help naming it).

Thanks

**Review by opr — DISMISSED** [review]
I've been looking at this a little more - a custom text field in the `contact` location does not show up in the "Account details" area, can you reproduce?

```php

		woocommerce_blocks_register_checkout_field(
			array(
				'id'       => 'second-plugin-namespace/gift-message-in-package',
				'label'    => 'Enter a gift message to include in the package',
				'location' => 'contact',
				'type'     => 'text',
			)
		);
```

**Comment by mikejolley** [comment]
@opr My Account > Account Details.

**Comment by opr** [comment]
@mikejolley 

Entered in the checkout block:
image

My account page:
image

**Comment by mikejolley** [comment]
@opr fixed in https://github.com/woocommerce/woocommerce/pull/43823/commits/237391aa832b61376daae5dc50ad89a0e96559cd

**Comment by mikejolley** [comment]
Been talking to @lsinger about the hook changes and what the best way forward is.

> So it seems that 21 plugins that we sell on [Woo.com](http://woo.com/) use that hook:

Majority of these extensions insert their own sections (fields, headings, that kind of thing). Moving the hook doesn't break this but it does change the position of these elements before or after the password section.

Personally I think it would make sense to have password as the last step and make sure all fields are before, but if we want to keep 100% bw compat we can create a new hook. What do you think @opr?

If we create a new hook it does give us one more thing to support when this gets changed to react.

**Comment by opr** [comment]
> Been talking to @lsinger about the hook changes and what the best way forward is.
> 
> > So it seems that 21 plugins that we sell on [Woo.com](http://woo.com/) use that hook:
> 
> Majority of these extensions insert their own sections (fields, headings, that kind of thing). Moving the hook doesn't break this but it does change the position of these elements before or after the password section.
> 
> Personally I think it would make sense to have password as the last step and make sure all fields are before, but if we want to keep 100% bw compat we can create a new hook. What do you think @opr?
> 
> If we create a new hook it does give us one more thing to support when this gets changed to react.

I think as long as we communicate this change, moving the hook location is fine. Maybe a blog post or GH discussion should go up after this is merged just so folks are aware.

**Comment by mikejolley** [comment]
@opr is there a process for this? In blocks repo we had the "needs dev note" label and added content in the description. I assume we've lost that here?

**Comment by opr** [comment]
> @opr is there a process for this? In blocks repo we had the "needs dev note" label and added content in the description. I assume we've lost that here?

I think we would post in the release thread pf59Ax-z6-p2

**Comment by lsinger** [comment]
> I think as long as we communicate this change, moving the hook location is fine.

I discussed this with Proton internally. I think the TL;DR is that this _seems_ OK to move and the usages that we found seem to be in favor of that theory. At the same time it's a change that makes some of feel like it has the potential to be a surprisingly problematic change because of an underestimation of how other extensions use our hooks (and other extension points). Therefore we wouldn't dismiss an alternative solution (like adding another hook instead of moving this one).

**Comment by mikejolley** [comment]
@opr ready for final review.

**Review by opr — APPROVED** [review]
Thanks Mike, it's working great now

**Pull request merged**
